### PR TITLE
Port and implement RestrictedInput TGUI component from upstream

### PR DESCRIPTION
**Pull request by Sovexe**
## About the PR 
* Ports the RestrictedInput component from upstream. Not quite a 1 to 1 as they've moved off of inferno. I had to refactor some bits like ReplaceAll which dont exist for us
* Implements the RestrictedInput componet in NumberInputModal
  * minor difference in arg names, upstream uses `round_value` where as we use `round_input`

## Why's this needed? 
* RestrictedInput brings with it autofocus and autoselect (which was the original purpose of this PR)

### CODE REVIEW

**Comment by mordent-goonstation** [comment]
LGTM. When we upgrade Inferno -> React would be worth looping back around to porting these back upstream so we're all in sync.

**Review by mordent-goonstation — APPROVED** [review]

**Comment by ZeWaka** [comment]
!merge_upstream

**Comment by ZeWaka** [comment]
!merge_upstream

**Comment by ZeWaka** [comment]
L+conflict+ratio

**Comment by ZeWaka** [comment]
!merge_upstream

**Comment by ZeWaka** [comment]
!merge_upstream

**Comment by ZeWaka** [comment]
what a shame, L

**Comment by ZeWaka** [comment]
!merge_upstream

**Comment by Sovexe** [comment]
converting back to a draft...there are some bugs to be addressed

**Comment by Sovexe** [comment]
Something I noticed with this (and with upstream) the input proc in dm still has an arg named `round_input` (`round_value` in upstream) but if its set to true you won't ever have anything to round. This is because the way the sanitization works numbers like `1.7` become `17`, the decimal is treated as a bad character to be stripped out. The round call does still occur in dm, but unless someone is exploiting shouldn't ever have a decimal to round. 

But that's how upstream has it so that's how it's been implemented here

**Review by ZeWaka — CHANGES_REQUESTED** [review]
zewaka needs to check licenses

File: tgui/packages/tgui/components/RestrictedInput.js [hunk 1]
```diff
@@ -0,0 +1,307 @@
+/**
+ * @file
+ * @copyright 2021
+ * @author jlsnow301 (https://github.com/jlsnow301)
+ * @license
```
Comment (ZeWaka) [inline tgui/packages/tgui/components/RestrictedInput.js:5]:
```suggestion
 * @license MIT
```

**Review by ZeWaka — APPROVED** [review]

**Comment by ZeWaka** [comment]
!merge_upstream

**Pull request merged**